File: nifti.py
# -*- coding: utf-8 -*-
"""
Simple script to read a 4D Nifti image using nibabel library and calculate the
sum of squared differences between the first 3D image and each of the other images.

@author: Javier Fernandez-Rial Portela
"""

import os
import numpy as np
import nibabel as nib
import timeit

# Setting file path and name (current directory by default)
PATH=""
file = os.path.join(PATH, 'sub-0003001_ses-1_task-rest_run-1_bold.nii')

# Loading Nifti image from nii file using nibabel library
img = nib.load(file)

# The 4 dimentions of the image, 3 spacial and one temporal
img.shape
# Output: (64, 64, 34, 200)

# The data type of each voxel is signed short int, of size 2 bytes
img.get_data_dtype()
# Output: dtype('<i2') (equivalent to numpy int16)

# Reading the header of the image
hdr = img.header
# Getting the vox_offset (extensions) of the image
hdr.get_data_offset()
# Output is 0, but that is false, the actual offset is 5376

# Getting the image data after the header and extensions
data = img.get_fdata()
# There are 200 3D images of dimensions (64,64,34)
data.shape
# Output: (64, 64, 34, 200)

# Getting the first 3D image
img1 = data[:,:,:,0]
img1.shape
# Output: (64, 64, 34)

# Converting first image to a 1 dimentional array of 139264 voxels
img1_flat = img1.flatten()
img1_flat.shape
# Output: (139264,)

# Getting the second 3D image
img2 = data[:,:,:,1]
img2_flat = img2.flatten()
img2_flat.shape
# Output: (139264,)

# Function to get the sum of squared differences between 2 3D images
# Params: two flattened 3D images
# Returns: ssd of the two images, an integer
def ssd(img1, img2):
    ssd = 0
    img_len = img1.shape[0]    
    for i in range(img_len):
        diff = img1[i] - img2[i]
        ssd += diff*diff
    return ssd

# Function to get a list of ssds between a target image and the rest 
# Params: the data from a 4D Nifti image
# Returns: a list of ssds, an integer list of size data.shape[3] - 1   
def get_ssd_list(data):
    target_img = data[:,:,:,0].flatten()
    ssd_list = []
    num_of_imgs = data.shape[3]
    for i in range(1,num_of_imgs):
        img = data[:,:,:,i].flatten()
        current_ssd = ssd(target_img,img)
        ssd_list.append(current_ssd)
    return ssd_list

# Getting a list of ssds from the data and printing the first and last ssd values
ssd_list = get_ssd_list(data)
print(ssd_list[0])
# Output: 36263554
print(ssd_list[-1])
# Output: 198750728

# Measuring the time it takes to run the ssd function (data.shape[3]-1)*10 times  
timeit.timeit('ssd(img1_flat, img2_flat)', number=1990, globals=globals())
# Output: 141.7s

# Measuring the time it takes to run the get_ssd_list function 10 times
timeit.timeit('get_ssd_list(data)', number=10, globals=globals())
# Output: 143.1s (average of 14.31s per call)


'''
For the sake of comparison, I show the initial versions of the previous functions
which were 4 times slower due to the use of numpy square or not flattening the 3D images
'''
# Slower version of the ssd function due to the use of numpy square
def ssd_slow(img1, img2):
    ssd = 0
    img_len = img1.shape[0]    
    for i in range(img_len):
        ssd += np.square(img1[i] - img2[i])
    return ssd

# Slower version of the get_ssd_list function due to not flattening the 3D images
def get_ssd_slow(data):
    ssd_list = []
    for i in range(1,data.shape[3]):
        ssd = 0
        for j in range(data.shape[2]):
            for k in range(data.shape[1]):
                for l in range(data.shape[0]):
                    diff = data[l,k,j,0] - data[l,k,j,i]
                    ssd += diff*diff
        ssd_list.append(ssd)
    return ssd_list

# Verifying that the output is the same
ssd_list_2 = get_ssd_slow(data)
print(ssd_list_2[0])
# Output: 36263554
print(ssd_list_2[-1])
# Output: 198750728

# Measuring the performance of the slower functions
timeit.timeit('ssd_slow(img1_flat, img2_flat)', number=1990, globals=globals())
# Output: 517.01s
timeit.timeit('get_ssd_slow(data)', number=10, globals=globals())
# Output: 546.52s (average of 54.65s per run, 4 times slower than get_ssd_list)

File: nifti.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

/*** signed short int has size 2 bytes */
#define BYTES_PER_READ 2
/*** 5376 offset + 352 nii_header_size = 5728 bytes */
#define NIFTI_HEADER_SIZE 5728
/*** product of the 3 dimensions of one 3D image (64*64*34) */
#define IMG_SIZE 139264
/*** there are 200 3D images on the nifti file indicated by its 4th dimension */
#define NUM_OF_IMAGES 200

struct Nifti{
    short *buffer;
};

/*** function to read the image data from a Nifti file with the previously defined characteristics */ 
struct Nifti read_img(char* filename){

    clock_t duration = clock();

    long size;
    size_t read;
    short *buffer;

    FILE *file = fopen(filename, "rb");

    if (!file) {
        fprintf(stderr, "Unable to open/create file\n");
        exit( 1 );
    }

    /*** Size of file */
    fseek(file, 0, SEEK_END);
    size = ftell(file) - NIFTI_HEADER_SIZE;

    if(!(buffer = malloc(size))) {
        fprintf(stderr, "Failed to allocate buffer memory\n");
        exit( 1 );
    }

    fseek(file, NIFTI_HEADER_SIZE, SEEK_SET);   
    read = fread(buffer, BYTES_PER_READ, size / BYTES_PER_READ, file);
    printf("I read %zu elements of size %d bytes (total bytes read: %zu)\n",
        read, BYTES_PER_READ, read * BYTES_PER_READ);

    struct Nifti img = {buffer};   
	fclose(file);

    duration = clock() - duration;
    double time_taken = ((double)duration)/CLOCKS_PER_SEC;

    printf("Elapsed read_img time: %fs\n", time_taken);

    return img;

}

/*** function to get an array of ssds between a target image and the rest */
long long *get_ssd_array( short* buffer, int num_of_images, int img_size ){

    clock_t duration = clock();
	
    long long *ssd_array;
    if(!(ssd_array = malloc(sizeof(long long)*(num_of_images - 1)))) {
        fprintf(stderr, "Failed to allocate ssd_array memory\n");
        exit( 1 );
    }
	
    long long ssd;
    short diff = 0;

    for( int j = 1; j < num_of_images; j = j + 1 ){
        ssd = 0;
	for( int  i = 0; i < img_size; i = i + 1 ){
            diff = buffer[i] - buffer[i + j*img_size];
            ssd += (long long)(diff*diff);
        }
        ssd_array[j-1] = ssd;
    }

    duration = clock() - duration;
    double time_taken = ((double)duration)/CLOCKS_PER_SEC;

    printf("Elapsed get_ssd_array time: %fs\n", time_taken);

    return ssd_array;

} 


int main(int argc, char *argv[])
{    
	char *nii_file;
	/*** set file path accordingly*/
	nii_file = "C:\\Users\\Javier\\Desktop\\BMB_1\\sub-0003001\\ses-1\\func\\sub-0003001_ses-1_task-rest_run-1_bold.nii";

	struct Nifti img = read_img(nii_file);

	long long *ssd_array = get_ssd_array(img.buffer, NUM_OF_IMAGES, IMG_SIZE);
	printf("ssd_first: %lld\n", ssd_array[0]);
	printf("ssd_last: %lld\n", ssd_array[NUM_OF_IMAGES - 2]);

    return 0;
}

/*** Sample output:
I read 27852800 elements of size 2 bytes (total bytes read: 55705600)
Elapsed read_img time: 0.421000s
Elapsed get_ssd_array time: 0.047000s
ssd_first: 36263554
ssd_last: 198750728
Press any key to continue... */
